include/eve/algo/container/soa_vector2.hpp

```diff
@@ -0,0 +1,372 @@
+    EVE_FORCEINLINE std::size_t size()  const noexcept { return storage.size_; }
+
+    //! Returns the number of elements that the container has currently allocated space for
+    EVE_FORCEINLINE std::size_t capacity()  const noexcept { return storage.capacity_; }
+
+    //! Requests the removal of unused capacity.
+    // EVE_FORCEINLINE void shrink_to_fit() noexcept
+    // {
+    //   kumi::for_each([](auto& m) { m.shrink_to_fit(); }, storage);
+    // }
+
+    //! Checks if the container has no elements
+    EVE_FORCEINLINE bool empty() const noexcept { return size() == 0ULL; }
+
+    //! @brief Increase the capacity of the vector to a value that's greater or equal to `n`.
+    //! If `n` is greater than the current capacity(), new storage is allocated and all iterators,
+    //! including the past-the-end iterator, and all references to the elements are invalidated.
+    //! Otherwise, the method does nothing.
+    //! @warning reserve() does not change the size of the vector.
+    //! @param n New capacity of the `eve::soa_vector`
+    void reserve(std::size_t n)
+    {
+      if(n > capacity())
+      {
+        auto const sz = size();
+        soa_vector that(n);
+        eve::algo::copy ( *this
+                        , algo::as_range(that.begin(), that.begin() + sz)
+                        );
+        this->operator=(std::move(that));
+        storage.size_ = sz;
+     }
+    }
+
+    //==============================================================================================
+    //! @}
+    //==============================================================================================
+
+    //==============================================================================================
+    //! @name Modifiers
+    //! @{
+    //==============================================================================================
+    //! @brief Clear the contents of the container.
+    void clear() { storage.size_ = 0; }
+
+    //! @brief Removes an element from the container.
+    //! Has the same invalidation semantics as std::vector.
+    //! end() iterator is not a valid pos.
+    iterator erase(const_iterator pos)
+    {
+      std::ptrdiff_t distance = pos - cbegin();
+      eve::algo::copy_backward( eve::algo::as_range(pos+1, cend())
+                              , eve::algo::as_range(begin()+distance, begin() + size() - 1)
+                              );
+      storage.size_--;
+      return begin() + distance;
+    }
+
+    //! @brief Removes the elements in the range [first, last)
+    //! Empty range is OK, does nothing
+    iterator erase(const_iterator f, const_iterator l)
+    {
+      std::ptrdiff_t distance_f = f - cbegin();
+      std::ptrdiff_t distance_l = l - cbegin();
+      std::ptrdiff_t sz = distance_l - distance_f;
+
+      eve::algo::copy_backward( eve::algo::as_range(l, cend())
+                              , eve::algo::as_range(begin() + distance_l - sz, end() - sz)
+                              );
```

Comment:
Let's extract the copy into a noinline function so that at least we are not duplicating between methods.